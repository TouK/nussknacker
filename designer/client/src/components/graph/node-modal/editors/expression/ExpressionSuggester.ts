import HttpService from "../../../../../http/HttpService";
import { ExpressionLang } from "./types";
import { VariableTypes } from "../../../../../types";

export type CaretPosition2d = { row: number; column: number };

export type ExpressionSuggestion = {
    methodName: string;
    refClazz: { display: string };
    fromClass: boolean;
    description?: string;
    parameters?: { name: string; refClazz: { display: string } }[];
};

export interface ExpressionSuggester {
    suggestionsFor(inputValue: string, caretPosition2d: CaretPosition2d): Promise<ExpressionSuggestion[]>;
}

export class BackendExpressionSuggester implements ExpressionSuggester {
    constructor(
        private language: ExpressionLang | string,
        private variableTypes: VariableTypes,
        private processingType: string,
        private httpService: typeof HttpService,
    ) {}

    suggestionsFor = (inputValue: string, caretPosition2d: CaretPosition2d): Promise<ExpressionSuggestion[]> =>
        this.httpService
            .getExpressionSuggestions(this.processingType, {
                expression: {
                    language: this.language,
                    expression: inputValue,
                },
                caretPosition2d,
                variableTypes: this.variableTypes,
            })
            .then((response) => response.data);
}
