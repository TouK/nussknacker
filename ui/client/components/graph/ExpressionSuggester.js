import _ from 'lodash';

export default class ExpressionSuggester {

  constructor(typesInformation, variables) {
    this._typesInformation = typesInformation
    this._variables = _.mapKeys(variables, (value, variableName) => {return `#${variableName}`})
  }

  suggestionsFor = (inputValue, caretPosition2d) => {
    const normalized = this._normalizeMultilineInputToSingleLine(inputValue, caretPosition2d)
    const lastExpressionPart = this._focusedLastExpressionPartWithoutMethodParens(normalized.normalizedInput, normalized.normalizedCaretPosition)
    const properties = this._alreadyTypedProperties(lastExpressionPart)
    const variablesIncludingSelectionOrProjection = this._getAllVariables(normalized);
    const focusedClazz = this._findRootClazz(properties, variablesIncludingSelectionOrProjection)
    return this._getSuggestions(lastExpressionPart, focusedClazz, variablesIncludingSelectionOrProjection)
  }

  _normalizeMultilineInputToSingleLine = (inputValue, caretPosition2d) => {
    const rows = inputValue.split("\n")
    const trimmedRows = _.map(rows, (row) => {
      const trimmedAtStartRow = _.dropWhile(row, (c) => c === " ").join("")
      return { trimmedAtStartRow: trimmedAtStartRow, trimmedCount: row.length - trimmedAtStartRow.length }
    })
    const beforeCaretInputLength = _.sum(_.map(_.take(trimmedRows, caretPosition2d.row), (row) => row.trimmedAtStartRow.length));
    const normalizedCaretPosition = caretPosition2d.column - trimmedRows[caretPosition2d.row].trimmedCount + beforeCaretInputLength
    const normalizedInput = _.map(trimmedRows, (row) => row.trimmedAtStartRow).join("")
    return {
      normalizedInput: normalizedInput,
      normalizedCaretPosition: normalizedCaretPosition
    }
  }

  _getSuggestions = (value, focusedClazz, variables) => {
    const variableNames = _.keys(variables)
    const variableAlreadySelected = _.some(variableNames, (variable) => { return _.includes(value, `${variable}.`) })
    const variableNotSelected = _.some(variableNames, (variable) => { return _.startsWith(variable.toLowerCase(), value.toLowerCase()) })
    let result = []
    if (variableAlreadySelected && focusedClazz) {
      const currentType = this._getTypeInfo(focusedClazz)
      const inputValue = this._justTypedProperty(value)
      const allowedMethodList = this._getAllowedMethods(currentType)
      result = inputValue.length === 0 ? allowedMethodList : this._filterSuggestionsForInput(allowedMethodList, inputValue)
    } else if (variableNotSelected && !_.isEmpty(value)) {
      const allVariablesWithClazzRefs = _.map(variables, (val, key) => {
        return {'methodName': key, 'refClazz': val}
      })
      result = this._filterSuggestionsForInput(allVariablesWithClazzRefs, value)
    }
    return new Promise((resolve, reject) => resolve(result))
  }

  _getAllowedMethods(currentType) {
    if (currentType.union != null) {
      const allMethods = _.flatMap(currentType.union, (subType) => this._getAllowedMethodsForClass(subType))
      // TODO: compute union of extracted methods types
      return _.uniqWith(allMethods, (typeA, typeB) => _.isEqual(typeA, typeB))
    } else {
      return this._getAllowedMethodsForClass(currentType)
    }
  }

  _getAllowedMethodsForClass(currentType) {
    return _.map(currentType.methods, (val, key) => {
      return {...val, methodName: key}
    })
  }

  _filterSuggestionsForInput = (variables, inputValue) => {
    return _.filter(variables, (variable) => {
      return _.includes(variable.methodName.toLowerCase(), inputValue.toLowerCase())
    })
  }

  _findRootClazz = (properties, variables) => {
    const variableName = properties[0]
    if (_.has(variables, variableName)) {
      const variableClazzName = _.get(variables, variableName);
      return _.reduce(_.tail(properties), (currentParentClazz, prop) => {
        const parentType = this._getTypeInfo(currentParentClazz)
        return this._extractMethod(parentType, prop)
      }, variableClazzName)
    } else {
      return null
    }
  }

  _extractMethod(type, prop) {
    if (type.union != null) {
      let foundedTypes = _.filter(_.map(type.union, (clazz) => this._extractMethodFromClass(clazz, prop)), i => i != null)
      // TODO: compute union of extracted methods types
      return _.first(foundedTypes) || {refClazzName: ''}
    } else {
      return this._extractMethodFromClass(type, prop) || {refClazzName: ''}
    }
  }

  _extractMethodFromClass(clazz, prop) {
    return _.get(clazz.methods, `${prop}.refClazz`)
  }

  _getTypeInfo = (type) => {
    if (type.union != null) {
      const unionOfTypeInfos = _.map(type.union, (clazz) => this._getTypeInfoFromClass(clazz))
      return {
        union: unionOfTypeInfos
      }
    } else {
      return this._getTypeInfoFromClass(type)
    }
  }

  _getTypeInfoFromClass = (clazz) => {
    const methodsFromInfo = this._getMethodsFromGlobalTypeInfo(clazz);
    const methodsFromFields = _.mapValues((clazz.fields || []), (field) => ({ refClazz: field }));
    const allMethods = _.merge(methodsFromFields, methodsFromInfo);

    return {
      ...clazz,
      methods: allMethods
    }
  }

  _getMethodsFromGlobalTypeInfo = (clazz) => {
    const foundData = _.find(this._typesInformation, { clazzName: { refClazzName: clazz.refClazzName }})
    return !_.isEmpty(foundData) ? foundData.methods : []
  }

  _focusedLastExpressionPartWithoutMethodParens = (expression, caretPosition) => {
    return this._lastExpressionPartWithoutMethodParens(this._currentlyFocusedExpressionPart(expression, caretPosition))
  }

  _currentlyFocusedExpressionPart = (value, caretPosition) => {
    return this._removeFinishedSelectionFromExpressionPart(value.slice(0, caretPosition))
  }

  //TODO: this does not handle map indices properly... e.g. #input.value.?[#this[""] > 4]
  _removeFinishedSelectionFromExpressionPart = (currentExpression) => {
    return currentExpression.replace(/\.\?\[[^\]]*]/g, "")
  }

  _lastExpressionPartWithoutMethodParens = (value) => {
    const withoutNestedParenthesis = value.substring(this._lastNonClosedParenthesisIndex(value) + 1, value.length);
    const valueCleaned = withoutNestedParenthesis.replace(/\(.*\)/, "");
    return _.isEmpty(value) ? "" : "#" + _.last(_.split(valueCleaned, '#'))
  };

  _lastNonClosedParenthesisIndex = (value) => {
    let nestingCounter = 0;
    for (let i = value.length - 1; i >= 0; i--) {
      if (value[i] === "(") nestingCounter -= 1;
      else if (value[i] === ")") nestingCounter += 1;

      if (nestingCounter < 0) return i;
    }
    return -1;
  };

  _justTypedProperty = (value) => {
    return _.last(this._dotSeparatedToProperties(value))
  }

  _alreadyTypedProperties = (value) => {
    return _.initial(this._dotSeparatedToProperties(value))
  }

  _dotSeparatedToProperties = (value) => {
    return _.split(value, ".")
  }

  _getAllVariables = (normalized) => {
    const thisClazz = this._findProjectionOrSelectionRootClazz(normalized)
    const data = thisClazz ? { '#this' : thisClazz } : {};
    return _.merge(data, this._variables)
  }

  _findProjectionOrSelectionRootClazz = (normalized) => {
    const currentProjectionOrSelection = this._findCurrentProjectionOrSelection(normalized)
    if (currentProjectionOrSelection) {
      const properties = this._alreadyTypedProperties(currentProjectionOrSelection)
      //TODO: currently we don't assume nested selections/projections
      const focusedClazz = this._findRootClazz(properties, this._variables)
      return (focusedClazz || {}).params ? focusedClazz.params[0] : null
    } else {
      return null
    }
  }

  _findCurrentProjectionOrSelection = (normalized) => {
    const input = normalized.normalizedInput;
    const caretPosition = normalized.normalizedCaretPosition;
    const currentPart = this._currentlyFocusedExpressionPart(input, caretPosition)
    //TODO: detect if it's *really* selection/projection (can be in quoted string, or method index??)
    const lastOpening = Math.max(currentPart.lastIndexOf("!["), currentPart.lastIndexOf("?["))
    const isInMiddleOfProjectionSelection = lastOpening > currentPart.indexOf("]")
    if (isInMiddleOfProjectionSelection) {
      const currentSelectionProjectionPart = currentPart.slice(0, lastOpening)
      //TODO: this won't handle former projections - but we don't support them now anyway...
      return this._lastExpressionPartWithoutMethodParens(currentSelectionProjectionPart)
    } else {
      return null;
    }

  }

}