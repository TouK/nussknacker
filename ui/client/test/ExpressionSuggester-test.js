import ExpressionSuggester from '../components/graph/node-modal/editors/expression/ExpressionSuggester'
import {map} from "lodash"

const typesInformation = [
  {
    "clazzName": {"refClazzName": "org.A"},
    "methods": {"fooString": {"refClazz": {"refClazzName": "java.lang.String"} }, "barB": {"refClazz": {"refClazzName": "org.B"} } }
  },
  {
    "clazzName": {"refClazzName": "org.B"},
    "methods": {"bazC": {"refClazz": {"refClazzName": "org.C"} } }
  },
  {
    "clazzName": {"refClazzName": "org.C"},
    "methods": {"quaxString": {"refClazz": {"refClazzName": "java.lang.String"} }}
  },
  {
    "clazzName": {"refClazzName": "org.AA"},
    "methods": {"fooString": {"refClazz": {"refClazzName": "java.lang.String"} }, "barB": {"refClazz": {"refClazzName": "org.C"} } }
  },
  {
    "clazzName": {"refClazzName": "org.WithList"},
    "methods": {"listField": {"refClazz": {"refClazzName": "java.util.List", params: [{refClazzName: "org.A"}]} }}
  },
  {
    "clazzName": {"refClazzName": "java.lang.String"},
    "methods": {"toUpperCase": {"refClazz": {"refClazzName": "java.lang.String"} }}
  },
  {
    "clazzName": {"refClazzName": "java.util.LocalDateTime"},
    "methods": {"isBefore": {"refClazz": {"refClazzName": "java.lang.Boolean"}, "params": {"name": "arg0", "refClazz": "java.util.LocalDateTime"}}}
  },
  {
    "clazzName": {"refClazzName": "org.Util"},
    "methods": {"now": {"refClazz": {"refClazzName": "java.util.LocalDateTime"}}}
  }
];

const variables = {
  "input": {refClazzName: "org.A"},
  "other": {refClazzName: "org.C"},
  "ANOTHER": {refClazzName: "org.A"},
  "dynamicMap": {refClazzName: "java.util.Map", fields: {'intField': {refClazzName: 'java.lang.Integer'}, 'aField': {refClazzName: "org.A"}} },
  "listVar": {refClazzName: "org.WithList" },
  "util": {refClazzName: "org.Util"},
  "union": {union: [
      {refClazzName: "org.A"},
      {refClazzName: "org.B"},
      {refClazzName: "org.AA"}]},
  "unionOfLists": {union: [
      {"refClazz": "java.util.List", params: [{refClazzName: "org.A"}]},
      {"refClazz": "java.util.List", params: [{refClazzName: "org.B"}]}
    ]},
  "dict": {dict: {
    id: "fooDict",
    valueType: {refClazzName: "org.A"}
  }}
};

const suggestionsFor = (inputValue, caretPosition2d, stubbedDictSuggestions) => {
  const _caretPosition2d = caretPosition2d ? caretPosition2d : { row: 0, column: inputValue.length }
  const stubService = {
    fetchDictLabelSuggestions(processingType, dictId, labelPattern) {
      return new Promise(resolve => resolve({ data: stubbedDictSuggestions }))
    }
  }
  const expressionSuggester = new ExpressionSuggester(typesInformation, variables, "fooProcessingType", stubService)
  return expressionSuggester.suggestionsFor(inputValue, _caretPosition2d)
}

describe("expression suggester", () => {
  it("should not suggest anything for empty input", () => {
    suggestionsFor("").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    })
  })

  it("should suggest all global variables if # specified", (done) => {
    suggestionsFor("#").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should filter global variables suggestions", (done) => {
    suggestionsFor("#ot").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should filter uppercase global variables suggestions", (done) => {
    suggestionsFor("#ANO").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest global variable", (done) => {
    suggestionsFor("#inpu").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest global variable methods", (done) => {
    suggestionsFor("#input.").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest dict variable methods", (done) => {
    let stubbedDictSuggestions = [
      { key: "one", label: "One" },
      { key: "two", label: "Two" }
    ]
    suggestionsFor("#dict.", null, stubbedDictSuggestions).then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })


  it("should suggest dict variable methods using indexer syntax", () => {
    let stubbedDictSuggestions = [
      { key: "sentence-with-spaces-and-dots", label: "Sentence with spaces and . dots" }
    ]
    const correctInputs = ["#dict['", "#dict['S", "#dict['Sentence w", "#dict['Sentence with spaces and . dots"]
    map(correctInputs, inputValue => {
      suggestionsFor(inputValue, null, stubbedDictSuggestions).then(suggestions => {
        expect(suggestions).toMatchSnapshot()
      })
    })
  })

  it("should suggest filtered global variable methods", (done) => {
    suggestionsFor("#input.fo").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest filtered global variable methods based not on beginning of the method", (done) => {
    suggestionsFor("#input.string").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest methods for object returned from method", (done) => {
    suggestionsFor("#input.barB.bazC.").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest methods for union objects", (done) => {
    suggestionsFor("#union.").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest methods for object returned from method from union objects", (done) => {
    suggestionsFor("#union.bazC.").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest in complex expression #1", (done) => {
    suggestionsFor("#input.foo + #input.barB.bazC.quax", {row: 0, column: "#input.foo".length }).then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest in complex expression #2", (done) => {
    suggestionsFor("#input.foo + #input.barB.bazC.quax").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest in complex expression #3", (done) => {
    suggestionsFor("#input.barB.bazC.quaxString.toUp").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })


  it("should not suggest anything if suggestion already applied with space at the end", (done) => {
    suggestionsFor("#input.fooString ").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest for invocations with method parameters #1", (done) => {
    suggestionsFor("#input.foo + #input.barB.bazC('1').quax").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest for invocations with method parameters #2", (done) => {
    suggestionsFor("#input.foo + #input.barB.bazC('1', #input.foo, 2).quax").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest for multiline code #1", (done) => {
    suggestionsFor("#input\n.fo", {row: 1, column: ".fo".length }).then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest for multiline code #2", (done) => {
    suggestionsFor("#input\n.barB\n.", {row: 2, column: ".".length }).then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest for multiline code #3", (done) => {
    suggestionsFor("#input\n.ba\n.barC", {row: 1, column: ".ba".length }).then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should omit whitespace formatting in suggest for multiline code #1", (done) => {
    suggestionsFor("#input\n  .ba", {row: 1, column: "  .ba".length }).then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should omit whitespace formatting in suggest for multiline code #2", (done) => {
    suggestionsFor("#input\n  .barB\n  .ba", {row: 2, column: "  .ba".length }).then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should omit whitespace formatting in suggest for multiline code #3", (done) => {
    suggestionsFor("#input\n  .ba\n  .bazC", {row: 1, column: "  .ba".length }).then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should omit whitespace formatting in suggest for multiline code #4", (done) => {
    suggestionsFor("#input\n  .barB.ba", {row: 1, column: "  .barB.ba".length }).then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should omit whitespace formatting in suggest for multiline code #5", (done) => {
    suggestionsFor("#input\n  .barB.bazC\n  .quaxString.", {row: 2, column: "  .quaxString.".length }).then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest field in typed map", (done) => {
    suggestionsFor("#dynamicMap.int").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest embedded field in typed map", (done) => {
    suggestionsFor("#dynamicMap.aField.f").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest #this fields in simple projection", (done) => {
    suggestionsFor("#listVar.listField.![#this.f]", {row: 0, column: "#listVar.listField.![#this.f".length }).then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest #this fields in projection on union of lists", (done) => {
    suggestionsFor("#unionOfLists.![#this.f]", {row: 0, column: "#unionOfLists.![#this.f".length }).then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should suggest #this fields in projection after selection", (done) => {
    suggestionsFor("#listVar.listField.?[#this == 'value'].![#this.f]", {row: 0, column: "#listVar.listField.?[#this == 'value'].![#this.f".length }).then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("handles negated parameters with projections and selections", (done) => {
    suggestionsFor("!#listVar.listField.?[#this == 'value'].![#this.f]", {row: 0, column: "!#listVar.listField.?[#this == 'value'].![#this.f".length }).then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should support nested method invocations", (done) => {
    suggestionsFor("#util.now(#other.quaxString.toUpperCase().)", {row: 0, column: "#util.now(#other.quaxString.toUpperCase().".length }).then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })

  it("should support safe navigation", (done) => {
    suggestionsFor("#input?.barB.bazC?.").then(suggestions => {
      expect(suggestions).toMatchSnapshot()
    }).then(done)
  })
})


describe("remove finished selections from input", () => {

  const expressionSuggester = new ExpressionSuggester(typesInformation, variables)

  it("leaves unfinished selection", () => {
    const original = "#input.one.two.?[#this."
    const removed = expressionSuggester._removeFinishedSelectionFromExpressionPart(original)
    expect(removed).toEqual(original)
  })

  it("leaves projections", () => {
    const original = "#input.one.two.![#this.value]"
    const removed = expressionSuggester._removeFinishedSelectionFromExpressionPart(original)
    expect(removed).toEqual(original)
  })

  it("removes one selection", () => {
    const original = "#input.one.two.?[#this.value == 1].three"
    const removed = expressionSuggester._removeFinishedSelectionFromExpressionPart(original)
    expect(removed).toEqual("#input.one.two.three")
  })

  it("removes many selections", () => {
    const original = "#input.one.two.?[#this.value == 1].three.?[#this.value == 2].four.?[#this.ala =="
    const removed = expressionSuggester._removeFinishedSelectionFromExpressionPart(original)
    expect(removed).toEqual("#input.one.two.three.four.?[#this.ala ==")
  })

})

describe("normalize multiline input", () => {

  const expressionSuggester = new ExpressionSuggester(typesInformation, variables)

  it("normalize multiline input #1", () => {
    const extracted = expressionSuggester._normalizeMultilineInputToSingleLine("#input\n  .barB.bazC\n  .quaxString.", {row: 1, column: "  .barB.bazC".length })
    expect(extracted).toEqual({
      normalizedInput: "#input.barB.bazC.quaxString.",
      normalizedCaretPosition: "#input.barB.bazC".length
    })
  })

  it("normalize multiline input #2", () => {
    const extracted = expressionSuggester._normalizeMultilineInputToSingleLine("#input\n  .barB.bazC\n  .quaxString.", {row: 2, column: "  .quaxString.".length })
    expect(extracted).toEqual({
      normalizedInput: "#input.barB.bazC.quaxString.",
      normalizedCaretPosition: "#input.barB.bazC.quaxString.".length
    })
  })

})
