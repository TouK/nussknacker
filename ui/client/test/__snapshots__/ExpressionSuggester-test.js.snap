// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expression suggester handles negated parameters with projections and selections 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should filter global variables suggestions 1`] = `
Array [
  Object {
    "methodName": "#other",
    "refClazz": Object {
      "refClazzName": "org.C",
    },
  },
]
`;

exports[`expression suggester should filter uppercase global variables suggestions 1`] = `
Array [
  Object {
    "methodName": "#ANOTHER",
    "refClazz": Object {
      "refClazzName": "org.A",
    },
  },
]
`;

exports[`expression suggester should not suggest anything for empty input 1`] = `Array []`;

exports[`expression suggester should not suggest anything if suggestion already applied with space at the end 1`] = `Array []`;

exports[`expression suggester should omit whitespace formatting in suggest for multiline code #1 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "barB",
    "refClazz": Object {
      "refClazzName": "org.B",
    },
  },
]
`;

exports[`expression suggester should omit whitespace formatting in suggest for multiline code #2 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "bazC",
    "refClazz": Object {
      "refClazzName": "org.C",
    },
  },
]
`;

exports[`expression suggester should omit whitespace formatting in suggest for multiline code #3 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "barB",
    "refClazz": Object {
      "refClazzName": "org.B",
    },
  },
]
`;

exports[`expression suggester should omit whitespace formatting in suggest for multiline code #4 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "bazC",
    "refClazz": Object {
      "refClazzName": "org.C",
    },
  },
]
`;

exports[`expression suggester should omit whitespace formatting in suggest for multiline code #5 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "toUpperCase",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest #this fields in projection after selection 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest #this fields in projection on union of lists 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest #this fields in simple projection 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest all global variables if # specified 1`] = `
Array [
  Object {
    "methodName": "#input",
    "refClazz": Object {
      "refClazzName": "org.A",
    },
  },
  Object {
    "methodName": "#other",
    "refClazz": Object {
      "refClazzName": "org.C",
    },
  },
  Object {
    "methodName": "#ANOTHER",
    "refClazz": Object {
      "refClazzName": "org.A",
    },
  },
  Object {
    "methodName": "#dynamicMap",
    "refClazz": Object {
      "fields": Object {
        "aField": Object {
          "refClazzName": "org.A",
        },
        "intField": Object {
          "refClazzName": "java.lang.Integer",
        },
      },
      "refClazzName": "java.util.Map",
    },
  },
  Object {
    "methodName": "#listVar",
    "refClazz": Object {
      "refClazzName": "org.WithList",
    },
  },
  Object {
    "methodName": "#util",
    "refClazz": Object {
      "refClazzName": "org.Util",
    },
  },
  Object {
    "methodName": "#union",
    "refClazz": Object {
      "union": Array [
        Object {
          "refClazzName": "org.A",
        },
        Object {
          "refClazzName": "org.B",
        },
        Object {
          "refClazzName": "org.AA",
        },
      ],
    },
  },
  Object {
    "methodName": "#unionOfLists",
    "refClazz": Object {
      "union": Array [
        Object {
          "params": Array [
            Object {
              "refClazzName": "org.A",
            },
          ],
          "refClazz": "java.util.List",
        },
        Object {
          "params": Array [
            Object {
              "refClazzName": "org.B",
            },
          ],
          "refClazz": "java.util.List",
        },
      ],
    },
  },
  Object {
    "methodName": "#dict",
    "refClazz": Object {
      "dict": Object {
        "id": "fooDict",
        "valueType": Object {
          "refClazzName": "org.A",
        },
      },
    },
  },
]
`;

exports[`expression suggester should suggest dict variable methods 1`] = `
Array [
  Object {
    "methodName": "One",
    "refClazz": Object {
      "refClazzName": "org.A",
    },
  },
  Object {
    "methodName": "Two",
    "refClazz": Object {
      "refClazzName": "org.A",
    },
  },
]
`;

exports[`expression suggester should suggest dict variable methods using indexer syntax 1`] = `
Array [
  Object {
    "methodName": "Sentence with spaces and . dots",
    "refClazz": Object {
      "refClazzName": "org.A",
    },
  },
]
`;

exports[`expression suggester should suggest dict variable methods using indexer syntax 2`] = `
Array [
  Object {
    "methodName": "Sentence with spaces and . dots",
    "refClazz": Object {
      "refClazzName": "org.A",
    },
  },
]
`;

exports[`expression suggester should suggest dict variable methods using indexer syntax 3`] = `
Array [
  Object {
    "methodName": "Sentence with spaces and . dots",
    "refClazz": Object {
      "refClazzName": "org.A",
    },
  },
]
`;

exports[`expression suggester should suggest dict variable methods using indexer syntax 4`] = `
Array [
  Object {
    "methodName": "Sentence with spaces and . dots",
    "refClazz": Object {
      "refClazzName": "org.A",
    },
  },
]
`;

exports[`expression suggester should suggest embedded field in typed map 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest field in typed map 1`] = `
Array [
  Object {
    "methodName": "intField",
    "refClazz": Object {
      "refClazzName": "java.lang.Integer",
    },
  },
]
`;

exports[`expression suggester should suggest filtered global variable methods 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest filtered global variable methods based not on beginning of the method 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest for invocations with method parameters #1 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "quaxString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest for invocations with method parameters #2 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "quaxString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest for multiline code #1 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest for multiline code #2 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "bazC",
    "refClazz": Object {
      "refClazzName": "org.C",
    },
  },
]
`;

exports[`expression suggester should suggest for multiline code #3 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "barB",
    "refClazz": Object {
      "refClazzName": "org.B",
    },
  },
]
`;

exports[`expression suggester should suggest global variable 1`] = `
Array [
  Object {
    "methodName": "#input",
    "refClazz": Object {
      "refClazzName": "org.A",
    },
  },
]
`;

exports[`expression suggester should suggest global variable methods 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
  Object {
    "fromClass": false,
    "methodName": "barB",
    "refClazz": Object {
      "refClazzName": "org.B",
    },
  },
]
`;

exports[`expression suggester should suggest in complex expression #1 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest in complex expression #2 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "quaxString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest in complex expression #3 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "toUpperCase",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest methods for object returned from method 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "quaxString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest methods for object returned from method from union objects 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "quaxString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest methods for union objects 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
  Object {
    "fromClass": false,
    "methodName": "barB",
    "refClazz": Object {
      "refClazzName": "org.B",
    },
  },
  Object {
    "fromClass": false,
    "methodName": "bazC",
    "refClazz": Object {
      "refClazzName": "org.C",
    },
  },
  Object {
    "fromClass": false,
    "methodName": "barB",
    "refClazz": Object {
      "refClazzName": "org.C",
    },
  },
]
`;

exports[`expression suggester should support nested method invocations 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "toUpperCase",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should support safe navigation 1`] = `
Array [
  Object {
    "fromClass": false,
    "methodName": "quaxString",
    "refClazz": Object {
      "refClazzName": "java.lang.String",
    },
  },
]
`;
